### PR TITLE
dkg: add naive sync protocol

### PR DESCRIPTION
**Pull request by dB2510**
Adds naive dkg sync protocol.

category: feature
ticket: #684

### CODE REVIEW

File: dkg/sync/server.go [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+	server.tcpNode.SetStreamHandler(ID, func(s network.Stream) {
+		defer s.Close()
+
+		buf := bufio.NewReader(s)
+		b := make([]byte, MsgSize)
+		_, err := buf.Read(b)
+		if err != nil {
+			log.Error(ctx, "Read client msg from s", err)
+			err = s.Reset()
+			log.Error(ctx, "Stream reset", err)
```
Comment (xenowits) [inline dkg/sync/server.go:56]: add return

File: dkg/sync/server.go [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+			resp.Error = "Invalid Signature"
+		}
+
+		resBytes, err := proto.Marshal(resp)
+		if err != nil {
+			log.Error(ctx, "Marshal server response", err)
+			err = s.Reset()
+			log.Error(ctx, "Stream reset", err)
+		}
+		_, err = s.Write(resBytes)
```
Comment (xenowits) [inline dkg/sync/server.go:81]: newline above this

**Review by xenowits — APPROVED** [review]

File: dkg/sync/server.go [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+	}
+
+	server.tcpNode.SetStreamHandler(ID, func(s network.Stream) {
+		defer s.Close()
+
+		buf := bufio.NewReader(s)
+		b := make([]byte, MsgSize)
+		_, err := buf.Read(b)
+		if err != nil {
+			log.Error(ctx, "Read client msg from s", err)
```
Comment (xenowits) [inline dkg/sync/server.go:54]:
```suggestion
			log.Error(ctx, "Read client msg from stream", err)
```

File: dkg/sync/client.go [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+
+type Client struct {
+	ctx          context.Context
+	onFailure    func()
+	tcpNode      host.Host
+	peer         p2p.Peer
+	serverStream network.Stream
+}
+
+// AwaitConnected blocks until the connection with the server has been established or returns an error.
```
[the remark this answers is not in the document]
Comment (dB2510) [reply dkg/sync/client.go:90]: yes

File: dkg/sync/server.go [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+	"google.golang.org/protobuf/proto"
+
+	"github.com/obolnetwork/charon/app/log"
+	pb "github.com/obolnetwork/charon/dkg/dkgpb/v1"
+	"github.com/obolnetwork/charon/p2p"
+)
+
+const (
+	ID      = "dkg_v1.0"
+	MsgSize = 112
```
[the remark this answers is not in the document]
Comment (dB2510) [reply dkg/sync/server.go:34]: MsgSync type has this size in bytes format after proto.Marshal. This variable is needed to read from stream.

File: dkg/sync/server.go [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+	"github.com/libp2p/go-libp2p-core/network"
+	"google.golang.org/protobuf/proto"
+
+	"github.com/obolnetwork/charon/app/log"
+	pb "github.com/obolnetwork/charon/dkg/dkgpb/v1"
+	"github.com/obolnetwork/charon/p2p"
+)
+
+const (
+	ID      = "dkg_v1.0"
```
[the remark this answers is not in the document]
Comment (dB2510) [reply dkg/sync/server.go:33]: sure

File: dkg/sync/server.go [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+)
+
+// NewServer registers a Stream Handler and returns a new Server instance.
+func NewServer(ctx context.Context, tcpNode host.Host, peers []p2p.Peer, hash []byte, onFailure func()) *Server {
+	server := &Server{
+		ctx:        ctx,
+		tcpNode:    tcpNode,
+		peers:      peers,
+		onFailure:  onFailure,
+		clientMsgs: make(chan *pb.MsgSync),
```
[the remark this answers is not in the document]
Comment (dB2510) [reply dkg/sync/server.go:44]: will remove this, was using for debugging in tests

File: dkg/sync/server.go (hunk 1 above)
Comment (dB2510) [reply dkg/sync/server.go:56]: sure

File: dkg/sync/client.go [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+	"github.com/obolnetwork/charon/app/log"
+	"github.com/obolnetwork/charon/app/z"
+	pb "github.com/obolnetwork/charon/dkg/dkgpb/v1"
+	"github.com/obolnetwork/charon/p2p"
+)
+
+// NewClient starts a goroutine that establishes a long lived connection and returns a new Client instance.
+func NewClient(ctx context.Context, tcpNode host.Host, server p2p.Peer, hash []byte, onFailure func(), ch chan *pb.MsgSyncResponse) Client {
+	go func() {
+		str, err := tcpNode.NewStream(ctx, server.ID, ID)
```
[the remark this answers is not in the document]
Comment (dB2510) [reply dkg/sync/client.go:36]: sure

**Pull request merged**